plugins/aver-hid/fu-aver-safeisp-device.c

```diff
@@ -0,0 +1,448 @@
+/*
+ * Copyright (C) 2023 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-aver-hid-firmware.h"
+#include "fu-aver-hid-struct.h"
+#include "fu-aver-safeisp-device.h"
+
+struct _FuAverSafeispDevice {
+	FuHidDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuAverSafeispDevice, fu_aver_safeisp_device, FU_TYPE_HID_DEVICE)
+
+#define FU_AVER_SAFEISP_DEVICE_TIMEOUT	     100000 /* ms */
+#define FU_AVER_SAFEISP_DEVICE_POLL_INTERVAL 5000   /* ms */
+
+enum { ISP_CX3, ISP_M12 };
+
+static gboolean
+fu_aver_safeisp_device_transfer(FuAverSafeispDevice *self,
+				GByteArray *req,
+				GByteArray *res,
+				GError **error)
+{
+	if (req != NULL) {
+		if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+					      req->data[0],
+					      req->data,
+					      req->len,
+					      FU_AVER_SAFEISP_DEVICE_TIMEOUT,
+					      FU_HID_DEVICE_FLAG_USE_INTERRUPT_TRANSFER,
+					      error)) {
+			g_prefix_error(error, "failed to send packet: ");
+			return FALSE;
+		}
+	}
+	if (res != NULL) {
+		if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+					      res->data[0],
+					      res->data,
+					      res->len,
+					      FU_AVER_SAFEISP_DEVICE_TIMEOUT,
+					      FU_HID_DEVICE_FLAG_USE_INTERRUPT_TRANSFER,
+					      error)) {
+			g_prefix_error(error, "failed to receive packet: ");
+			return FALSE;
+		}
+		g_debug("custom-isp-cmd: %s [0x%x]",
+			fu_aver_safeisp_custom_cmd_to_string(
+			    fu_struct_aver_safeisp_res_get_custom_cmd(res)),
+			fu_struct_aver_safeisp_res_get_custom_cmd(res));
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_poll(FuDevice *device, GError **error)
+{
+	FuAverSafeispDevice *self = FU_AVER_SAFEISP_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GByteArray) req = fu_struct_aver_safeisp_req_new();
+	g_autoptr(GByteArray) res = fu_struct_aver_safeisp_res_new();
+
+	locker = fu_device_locker_new(device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	fu_struct_aver_safeisp_req_set_custom_cmd(req, FU_AVER_SAFEISP_CUSTOM_CMD_GET_VERSION);
+	if (!fu_aver_safeisp_device_transfer(self, req, res, error))
+		return FALSE;
```

Comment:
I'm a bit confused why you have to poll the version number again and again.

Reply:
CAM340+ does not have a specific request command for status response. Therefore, we utilize the version command to verify the existence of CAM340+.